File: backend/app/models/dlp.py
from datetime import datetime
from uuid import UUID, uuid4

from sqlalchemy import Boolean, DateTime, ForeignKey, Index, Integer, String, Text, UniqueConstraint
from sqlalchemy.orm import Mapped, mapped_column
from sqlalchemy.sql import func

from app.database import Base


class DLPConfig(Base):
    __tablename__ = "dlp_configs"

    id: Mapped[UUID] = mapped_column(primary_key=True, default=uuid4)
    tenant_id: Mapped[UUID] = mapped_column(ForeignKey("tenants.id", ondelete="CASCADE"), nullable=False, unique=True)
    enabled: Mapped[bool] = mapped_column(Boolean, nullable=False, default=True)
    compliance_mode: Mapped[bool] = mapped_column(Boolean, nullable=False, default=False)
    incident_threshold_count: Mapped[int] = mapped_column(Integer, nullable=False, default=5)
    incident_threshold_hours: Mapped[int] = mapped_column(Integer, nullable=False, default=24)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())
    updated_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now(), onupdate=func.now())


class DLPRule(Base):
    __tablename__ = "dlp_rules"
    __table_args__ = (
        UniqueConstraint("tenant_id", "rule_key", name="uq_dlp_rules_tenant_key"),
        Index("ix_dlp_rules_tenant_id", "tenant_id"),
    )

    id: Mapped[UUID] = mapped_column(primary_key=True, default=uuid4)
    tenant_id: Mapped[UUID] = mapped_column(ForeignKey("tenants.id", ondelete="CASCADE"), nullable=False)
    rule_key: Mapped[str] = mapped_column(String(64), nullable=False)
    rule_name: Mapped[str] = mapped_column(String(100), nullable=False)
    description: Mapped[str | None] = mapped_column(String(255), nullable=True)
    category: Mapped[str] = mapped_column(String(32), nullable=False)
    action: Mapped[str] = mapped_column(String(8), nullable=False, default="block")
    is_enabled: Mapped[bool] = mapped_column(Boolean, nullable=False, default=True)
    is_builtin: Mapped[bool] = mapped_column(Boolean, nullable=False, default=True)
    pattern: Mapped[str | None] = mapped_column(Text, nullable=True)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())
    updated_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now(), onupdate=func.now())


class DLPIncident(Base):
    __tablename__ = "dlp_incidents"
    __table_args__ = (
        Index("ix_dlp_incidents_tenant_id", "tenant_id"),
        Index("ix_dlp_incidents_created_at", "created_at"),
    )

    id: Mapped[UUID] = mapped_column(primary_key=True, default=uuid4)
    tenant_id: Mapped[UUID] = mapped_column(ForeignKey("tenants.id", ondelete="CASCADE"), nullable=False)
    user_id: Mapped[UUID | None] = mapped_column(ForeignKey("users.id", ondelete="SET NULL"), nullable=True)
    pii_type: Mapped[str] = mapped_column(String(64), nullable=False)
    action_taken: Mapped[str] = mapped_column(String(8), nullable=False)
    source: Mapped[str] = mapped_column(String(32), nullable=False, default="chat")
    ip_address: Mapped[str | None] = mapped_column(String(45), nullable=True)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())

File: frontend/src/api/dlp.ts
import apiClient from "./client";

export interface DLPConfig {
  id: string;
  tenant_id: string;
  enabled: boolean;
  compliance_mode: boolean;
  incident_threshold_count: number;
  incident_threshold_hours: number;
}

export interface DLPRule {
  id: string;
  rule_key: string;
  rule_name: string;
  description: string | null;
  category: "pii_br" | "credentials" | "infra_mssp" | "custom";
  action: "block" | "warn";
  is_enabled: boolean;
  is_builtin: boolean;
  pattern: string | null;
}

export interface DLPIncident {
  id: string;
  pii_type: string;
  action_taken: "block" | "warn";
  source: string;
  ip_address: string | null;
  user_id: string | null;
  created_at: string;
}

export interface DLPRuleCreate {
  rule_key: string;
  rule_name: string;
  description?: string;
  category?: string;
  action: "block" | "warn";
  pattern: string;
}

const qs = (tenantId?: string) => (tenantId ? `?tenant_id=${tenantId}` : "");

export const dlpApi = {
  getConfig: (tenantId?: string): Promise<DLPConfig> =>
    apiClient.get(`/dlp/config${qs(tenantId)}`).then((r) => r.data),

  updateConfig: (
    data: Partial<Pick<DLPConfig, "enabled" | "compliance_mode" | "incident_threshold_count" | "incident_threshold_hours">>,
    tenantId?: string,
  ): Promise<DLPConfig> =>
    apiClient.put(`/dlp/config${qs(tenantId)}`, data).then((r) => r.data),

  listRules: (tenantId?: string): Promise<DLPRule[]> =>
    apiClient.get(`/dlp/rules${qs(tenantId)}`).then((r) => r.data),

  updateRule: (ruleId: string, data: { action?: "block" | "warn"; is_enabled?: boolean }, tenantId?: string): Promise<DLPRule> =>
    apiClient.put(`/dlp/rules/${ruleId}${qs(tenantId)}`, data).then((r) => r.data),

  createRule: (data: DLPRuleCreate, tenantId?: string): Promise<DLPRule> =>
    apiClient.post(`/dlp/rules${qs(tenantId)}`, data).then((r) => r.data),

  deleteRule: (ruleId: string, tenantId?: string): Promise<void> =>
    apiClient.delete(`/dlp/rules/${ruleId}${qs(tenantId)}`).then(() => undefined),

  listIncidents: (tenantId?: string, limit = 50): Promise<DLPIncident[]> =>
    apiClient.get(`/dlp/incidents${qs(tenantId)}&limit=${limit}`).then((r) => r.data),
};
